src/api/logs.rs

```diff
@@ -0,0 +1,329 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+use std::sync::{Arc, RwLock};
+
+use futures::future::Either;
+use futures::sink::Sink;
+use futures::{future, Future, Stream};
+use hyper::{header, Body, Chunk, Request, Response};
+use tokio::sync::mpsc::unbounded_channel;
+
+use crate::api::{ListResponse, SafeOutput, ViewSet};
+use crate::config::{Config, Log};
+use crate::http::{return_400, return_404, return_500, ResponseFuture};
+use crate::storage::{delete_object_metabucket, put_object_metabucket};
+
+pub struct ApiLogs {
+    config: Arc<RwLock<Config>>,
+}
+
+impl SafeOutput for Log {
+    // Log has nothing to hide
+    fn safe(&mut self) {}
+}
+
+impl ApiLogs {
+    pub fn new(cfg: Arc<RwLock<Config>>) -> ApiLogs {
+        ApiLogs { config: cfg }
+    }
+
+    // Parses the log from the create body; returns error response in
+    // case it is not valid.
+    fn parse_create_body(
+        entire_body: Vec<u8>,
+        cfg: Arc<RwLock<Config>>,
+    ) -> Result<Log, Response<Body>> {
+        let payload = String::from_utf8(entire_body)
+            .map_err(|_| return_400("Could not understand request"))?;
+
+        let log: Log =
+            serde_json::from_str(&payload).map_err(|_| return_400("Could not parse request"))?;
+
+        // Validate Commit Window
+        if log.commit_window == "" {
+            return Err(return_400("Commit window key cannot be empty."));
+        }
+        if !log.commit_window.ends_with("s") && !log.commit_window.ends_with("m") {
+            return Err(return_400(
+                "Commit window must be specified in either seconds `5s` or minutes `1m`",
+            ));
+        }
+
+        // if the commit window parses to 0 and the value is not 0, 0s or 0m, it's an invalid window
+        let parsed_window = Config::commit_window_to_seconds(&log.commit_window);
+        if log.commit_window != "0"
+            && log.commit_window != "0s"
+            && log.commit_window != "0m"
+            && parsed_window == 0
+        {
+            return Err(return_400("Commit window is invalid"));
+        }
+
+        let cfg_read = cfg.read().unwrap();
+        // validate the datastores
+        for ds_name in &log.datastores {
+            if cfg_read.datastore.contains_key(ds_name) == false {
+                return Err(return_400(&format!(
+                    "{} is an invalid datastore name",
+                    &ds_name
+                )));
+            }
+        }
+
+        // Validate name
+
+        if let Some(lg_name) = &log.name {
```

Comment:
In the Log data type why is the name an Option? I think it's always needed. We just need to validate that it is not an empty string.

Reply:
good point, that's a large change dropping the `Option<>` from log.name, let's do that on a separate PR since that will touch many files